src/extra/components/adq.py

```diff
@@ -0,0 +1,1318 @@
+                unstacked_data.shape[1]) * pulse_duration
+        else:
+            raise ValueError(f'invalid pulse dimension `{pulse_dim}`')
+
+        coords.update(self._build_sample_coords(unstacked_data))
+
+        import xarray as xr
+        return xr.DataArray(unstacked_data, coords=coords)
+
+    def find_edges(self, data, edge_func=None, max_edges=50, parallel=None,
+                   **edge_kw):
+        """Find signal edges.
+
+        In some cases, not the raw data itself may be of interest but
+        the location (and amplitude) of certain signals in the raw data.
+        One such example is time-of-flight spectroscopy in counting
+        mode, where individual charged particles impact a detection
+        surface and leave a fast signal on the digitizer trace. Fast
+        timing discriminators allow to robustly determine the position
+        of such signals.
+
+        By default, it uses the
+        [dynamic leading discriminator](extra.signal.dled) from the
+        [extra.signal](extra.signal) package, but other from this
+        package or entirely custom functions may be used as well. The
+        required signatures must include three keyword arguments
+        `signal`, `edges` and `amplitudes` corresponding to those from
+        [extra.signal.dled](extra.signal.dled). The default edge finding
+        method requires the `threshold` parameter to be passed as
+        keyword argument.
+
+        The processing is parallelized via `pasha`.
+
+        Args:
+            data (ArrayLike): Input data to find edges on.
+            edge_func (Callable, optional): Edge finding method to run
+                on each train trace, extra.signal.dled by default.
+            max_edges (int, optional): Maximal number of edges per
+                train, 50 by default.
+            parallel (int or None, optional): Nunmber of parallel
+                processes to use, by default 10 or a quarter of all cores
+                whichever is lower. Any non-positive value or 1 disable
+                parallelization.
+
+            Any further keyword arguments are passed to the edge finding
+            method.
+
+        Returns:
+            result (pandas.DataFrame): Edge positions and pulse heights.
+        """
+
+        edges, amplitudes = self.find_edge_array(
+            data, False, False, edge_func, max_edges, parallel, **edge_kw)
+
+        if data.ndim > 2:
+            index = pd.MultiIndex.from_product(
+                [np.arange(x) for x in data.shape[:-1]],
+                names=[f'dim_{i}' for i in range(data.ndim - 1)])
+        else:
+            index = pd.Index(np.arange(data.shape[0]), name='dim_0')
+
+        return self._shape_edges(edges, amplitudes, index)
+
+    def find_edge_array(self, data, labelled=True, squeeze_edges=True,
+                        edge_func=None, max_edges=50, parallel=None,
+                        **edge_kw):
+        """Find signal edges as sparse array.
```

Comment:
To my mind, a 'sparse array' means a specialised data structure for representing a large array where only a small fraction of values are filled in, using less memory than the normal ('dense') array representation. As in the [scipy.sparse](https://docs.scipy.org/doc/scipy/reference/sparse.html) package, for instance.

From the code, I don't think that's what you mean here? I think you're returning something conceptually like what I'd call a 'ragged' array, i.e. the number of entries along one axis varies, but the entries that are there are always at the start of the axis. I'm not sure I'd use that term either for a regular array with NaNs in, though. :thinking:

Reply:
TIL, I was not sure how to call them. It is sparse in terms of contained information, but I'll go with ragged.